lib/src/footer.rs

```diff
@@ -0,0 +1,132 @@
+// Copyright 2024 The Jujutsu Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+// https://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+//! Parsing footer lines from commit messages.
+
+/// A key-value pair representing a footer line in a commit message, of the
+/// form `Key: Value`.
+#[derive(Debug, PartialEq, Clone)]
+pub struct FooterEntry(pub String, pub String);
+
+/// Parse the footer lines from a commit message; these are simple key-value
+/// pairs, separated by a colon, describing extra information in a commit
+/// message; an example is the following:
+///
+/// ```text
+/// chore: fix bug 1234
+///
+/// Lorem ipsum dolor sit amet, consectetur adipiscing elit, sed do eiusmod
+/// tempor incididunt ut labore et dolore magna aliqua.
+///
+/// Co-authored-by: Alice <alice@example.com>
+/// Co-authored-by: Bob <bob@example.com>
+/// Reviewed-by: Charlie <charlie@example.com>
+/// Change-Id: I1234567890abcdef1234567890abcdef12345678
+/// ```
+///
+/// In this case, there are four footer lines: two `Co-authored-by` lines, one
+/// `Reviewed-by` line, and one `Change-Id` line.
+pub fn get_footer_lines(body: &str) -> Vec<FooterEntry> {
```

Comment:
It seems like the code would be made simpler by making this into a `Map<String, Vec<String>>`.

Reply:
A drive-by comment, but a `Map<String, Vec<String>>` would lose any ordering of footers, which _may_ not be desirable if this is to be used in some form of commit message rewrite.

Maybe something like https://github.com/indexmap-rs/indexmap which preserves insertion order?

Reply:
> A drive-by comment, but a `Map<String, Vec<String>>` would lose any ordering of footers, which _may_ not be desirable if this is to be used in some form of commit message rewrite.

This is correct, but the reason I suggested a Map was because the map is never read in a way that makes ordering preserved being relevant.

The only operations that are performed on it are:
* Read the key "ChangeId"  as an `Option<Vec<String>>`
* Check if the footer is entirely empty.

> Maybe something like https://github.com/indexmap-rs/indexmap which preserves insertion order?

If you had an `IndexMap<String, Vec<String>>`, then you can't preserve the following footer ordering, which I assume is valid. If you do want to preserve the order, then I think the current approach is the correct one:

```
Foo: a
Bar: b
Foo: c
```

Reply:
> If you had an `IndexMap<String, Vec<String>>`, then you can't preserve the following footer ordering,

Correct. I think I suggested `Vec<(String, String)>` for that reason. It can also return `Iterator` to let caller decide the data type.

Reply:
Yes, I followed Yuya's suggestion to revert this to a `Vec`; it was originally a Map, but I decided that trying to preserve the input ordering was the best and least surprising user outcome for a user.